HB/common/synthesis.elpi

```diff
@@ -1,93 +1,126 @@
 
+% Given TheType and a factory instance (on it), makes all the mixins provided by
+% the factory available for inference.
 pred under-mixin-src-from-factory.do! i:term, i:term, i:list prop.
 under-mixin-src-from-factory.do! TheType TheFactory LP :-
-  std.under.do! (under-mixin-src-from-factory.then TheType TheFactory) LP.
-
-pred under-mixin-src-from-factories.then i:term, i:list term, i:(term -> prop), o:term.
-under-mixin-src-from-factories.then TheType Factories P X :-
-  std.map Factories (mixin-srcs TheType) MLL,
-  std.flatten MLL ML,
-  ML => P X.
-
-pred under-mixin-src-from-factories.do! i:term, i:list term, i:list prop.
-under-mixin-src-from-factories.do! TheType Factories LP :-
-  std.under.do! (under-mixin-src-from-factories.then TheType Factories) LP.
-
+  private.factory-instance->new-mixins [] TheFactory ML,
+  std.map ML (m\c\ c = mixin-src TheType m TheFactory) MLClauses,
+  MLClauses => std.do! LP.
+
+% Given TheType and a factory instance (on it), builds all the *new* mixins
+% provided by the factory available for and passes them to the given
+% continuation
+pred under-new-mixin-src-from-factory.do! i:term, i:term, i:(list mixinname -> prop).
+under-new-mixin-src-from-factory.do! TheType TheFactory LP :-
+  findall-mixin-src TheType OldMixins,
+  private.factory-instance->new-mixins OldMixins TheFactory NewML,
+  std.map NewML (m\c\ c = mixin-src TheType m TheFactory) NewMLClauses,
+  NewMLClauses => std.do! [ LP NewML ].
+
+% [under-mixins.then MLwP Pred F] states that F has shape
+%   fun p_1 .. p_k T,
+%      (m_0 : M_0 ..p.. T) .. (m_n : M_n ..p.. T m_i0 .. m_ik) =>
+%      Body m_0 .. m_n
+% where  MLwP contains M_0, .., M_n (under p_1 .. p_k)
+%   and  Body is such that [..,mixin-src T M_i m_i,..] => Pred Body
+%   and  ..p.. is a list of terms built using p_1 .. p_k and T
 pred under-mixins.then i:list (w-args mixinname),
     i:(name -> term -> (term -> A) -> A -> prop),
     i:(A -> prop), o:A.
 under-mixins.then [] _ Pred Body :- !, Pred Body.
-under-mixins.then [triple M Args T|ML] Mixin Pred Out :- std.do! [
-  mgref->term Args T M MTy,
+under-mixins.then [triple M Args T|ML] MkFun Pred Out :- std.do! [
+  infer-all-mixin-args Args T M MTy,
   (@pi-decl `m` MTy m\ mixin-src T M m =>
-    under-mixins.then ML Mixin Pred (Body m)),
-  Mixin `m` MTy Body Out,
+    under-mixins.then ML MkFun Pred (Body m)),
+  MkFun `m` MTy Body Out,
 ].
 
-% [mk-mixin-fun.then MLwP Pred F] states that F has shape
-%   fun p_1 .. p_k T,
-%      (m_0 : M_0 ..p.. T) .. (m_n : M_n ..p.. T m_i0 .. m_ik) =>
-%      Body m_0 .. m_n
-% where  MLwP contains M_0, .., M_n (under p_1 .. p_k)
-%   and  Body is such that [..,mixin-src T M_i m_i,..] => Pred Body
-%   and  ..p.. is a list of terms built using p_1 .. p_k and T
-pred mk-mixin-fun.then i:list-w-params mixinname, i:(list term -> term -> term -> prop), o:term.
-mk-mixin-fun.then L P Out :- !,
-  w-params.then L mk-fun mk-fun
-    (p\ t\ ml\ under-mixins.then ml mk-fun (P p t)) Out.
```

Comment:
I need this one [here](https://github.com/math-comp/hierarchy-builder/pull/171/files#diff-6f6b8f9893a3436966238c846d42ef4a44be696c850f37a3d270ea1abd08e2aaR257)

Reply:
This is 99% equal to under-mixins.then which I left and which kept this comment

Reply:
I'm suggesting to inline it

Reply:
see the change to `declare-id-builder`, it is more verbose but also simpler IMO

Reply:
I saw the in-lining, and I would really like to keep the non-inlined version. I actually meant to use it a couple more times in the future.

Reply:
> see the change to `declare-id-builder`, it is more verbose but also simpler IMO

It's not simpler to me...  you need `aux` and `aux2` while one of them is actually generic

Reply:
The name was bad though...

Reply:
`w-params.then MLwP mk-fun mk-fun` is the only part we really need to inline (the only difference w.r.t. under-mixins.then).
The choice of not using anonymous clauses is orthogonal, and makes things more verbose but easier for a beginner IMO.

Reply:
I mean, 4 lines to share 1 (in 2 places) does not seem a good ratio to me.

In Coq we have gazilions of `let f x = g x + epsilon` and I find it always hard to follow, you have to track f down just to discover you don't learn much from it since the important thing is actually g.

Reply:
> `w-params.then MLwP mk-fun mk-fun` is the only part we really need to inline (the only difference w.r.t. under-mixins.then).

`mk-mixin-fun.then` is the composition of `w-params.then` with `under-mixins.then`, and `mk|fun` is then one of the only sound choices to fill the rest... As a high-level abstraction it makes sense and it can be used to factor at least two functions, which is enough for me. It will be the basis of custom `pack` notations as well.


> I mean, 4 lines to share 1 (in 2 places) does not seem a good ratio to me.

This is not the real ratio.
This diff removes 8 lines and replaces them with 7.

https://github.com/math-comp/hierarchy-builder/pull/174/files#diff-6f6b8f9893a3436966238c846d42ef4a44be696c850f37a3d270ea1abd08e2aaL71-R80

should in fact be replaced by, which removes 5 lines and replaces them with 3.
```
mk-mixin-fun.then {gref-deps GR} declare-id-builder.aux GR
```
```
declare-id-builder.aux GR Params TheType (fun `x` Ty x\ x) :- 
  synthesis.infer-all-mixin-args Params TheType GR Ty.
```

The abstraction `mk-mixin-fun.then` saves space too.
By removing it you made [this](https://github.com/math-comp/hierarchy-builder/pull/174/files#diff-6f6b8f9893a3436966238c846d42ef4a44be696c850f37a3d270ea1abd08e2aaR76) a three line specific piece of code, while `mk-mixin-fun.then` did it in a generic way...

Reply:
Maybe this would be clearer with a commit...

Reply:
See https://github.com/math-comp/hierarchy-builder/commit/b6c13c9fa4b1d73de628c34439f7db4eddd9ddc7

Reply:
Nevermind. can you push you commit here?

Reply:
done